File: Redis350/publisher-module/src/main/java/by/rest/publisher/service/TagService.java
package by.rest.publisher.service;

import by.rest.publisher.domain.Tag;
import by.rest.publisher.dto.TagRequestTo;
import by.rest.publisher.dto.TagResponseTo;
import by.rest.publisher.exception.ApiException;
import by.rest.publisher.mapper.TagMapper;
import by.rest.publisher.repository.TagRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class TagService {
    
    private final TagRepository tagRepository;
    private final TagMapper tagMapper;
    
    public TagService(TagRepository tagRepository, TagMapper tagMapper) {
        this.tagRepository = tagRepository;
        this.tagMapper = tagMapper;
    }
    
    public TagResponseTo create(TagRequestTo request) {
        validateTagRequest(request);
        
        // Проверка уникальности имени тега
        Optional<Tag> existingTag = tagRepository.findByName(request.getName());
        if (existingTag.isPresent()) {
            throw new ApiException(400, "40020", "Tag with name '" + request.getName() + "' already exists");
        }
        
        Tag tag = tagMapper.toEntity(request);
        tag = tagRepository.save(tag);
        return tagMapper.toResponse(tag);
    }
    
    @Transactional(readOnly = true)
    public List<TagResponseTo> getAll() {
        return tagRepository.findAll().stream()
                .map(tagMapper::toResponse)
                .toList();
    }
    
    @Transactional(readOnly = true)
    public TagResponseTo getById(Long id) {
        Tag tag = tagRepository.findById(id)
                .orElseThrow(() -> new ApiException(404, "40402", "Tag not found with id: " + id));
        return tagMapper.toResponse(tag);
    }
    
    public TagResponseTo update(Long id, TagRequestTo request) {
        validateTagRequest(request);
        
        Tag tag = tagRepository.findById(id)
                .orElseThrow(() -> new ApiException(404, "40402", "Tag not found with id: " + id));
        
        // Проверка уникальности имени тега (кроме текущего тега)
        if (!tag.getName().equals(request.getName())) {
            Optional<Tag> existingTag = tagRepository.findByName(request.getName());
            if (existingTag.isPresent()) {
                throw new ApiException(400, "40020", "Tag with name '" + request.getName() + "' already exists");
            }
        }
        
        tag.setName(request.getName());
        tag = tagRepository.save(tag);
        return tagMapper.toResponse(tag);
    }
    
    public void delete(Long id) {
        if (!tagRepository.existsById(id)) {
            throw new ApiException(404, "40402", "Tag not found with id: " + id);
        }
        tagRepository.deleteById(id);
    }
    
    @Transactional(readOnly = true)
    public Optional<TagResponseTo> findByName(String name) {
        return tagRepository.findByName(name)
                .map(tagMapper::toResponse);
    }
    
    private void validateTagRequest(TagRequestTo request) {
        if (request.getName() == null || request.getName().trim().isEmpty()) {
            throw new ApiException(400, "40021", "Tag name cannot be empty");
        }
        if (request.getName().length() < 2 || request.getName().length() > 32) {
            throw new ApiException(400, "40022", "Tag name must be between 2 and 32 characters");
        }
    }
}

File: 351002/Melnikov/publisher/services/tag_service.py
from ..common.errors import ConflictError, NotFoundError
from ..common.validation import validate_length
from ..dto.tag_dto import TagRequestTo, TagResponseTo
from ..models.tag import Tag
from ..repositories.registry import tag_repository


class TagService:
    def create(self, dto: TagRequestTo) -> TagResponseTo:
        name = validate_length("name", dto.name, 2, 32, 21)

        existing = [t for t in tag_repository.get_all() if t.name == name]
        if existing:
            raise ConflictError("Tag with this name already exists", 2)

        tag = Tag(id=0, name=name)
        created = tag_repository.create(tag)
        return self._to_response(created)

    def get_all(self) -> list[TagResponseTo]:
        return [self._to_response(tag) for tag in tag_repository.get_all()]

    def get_by_id(self, tag_id: int) -> TagResponseTo:
        tag = tag_repository.get_by_id(tag_id)
        if not tag:
            raise NotFoundError("Tag not found", 3)
        return self._to_response(tag)

    def update(self, tag_id: int, dto: TagRequestTo) -> TagResponseTo:
        tag = tag_repository.get_by_id(tag_id)
        if not tag:
            raise NotFoundError("Tag not found", 3)

        name = validate_length("name", dto.name, 2, 32, 21)

        for item in tag_repository.get_all():
            if item.name == name and item.id != tag_id:
                raise ConflictError("Tag with this name already exists", 2)

        tag.name = name
        updated = tag_repository.update(tag)
        return self._to_response(updated)

    def delete(self, tag_id: int) -> None:
        deleted = tag_repository.delete(tag_id)
        if not deleted:
            raise NotFoundError("Tag not found", 3)

    def get_many_by_ids(self, ids: list[int]) -> list[TagResponseTo]:
        result = []
        for tag_id in ids:
            tag = tag_repository.get_by_id(tag_id)
            if tag:
                result.append(self._to_response(tag))
        return result

    @staticmethod
    def _to_response(tag: Tag) -> TagResponseTo:
        return TagResponseTo(
            id=tag.id,
            name=tag.name
        )